src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -45,33 +60,213 @@ static NSStringCompareOptions ConvertFromCompareOptionsToNSStringCompareOptions(
+
+    if (error != nil)
+        return source;
+
+    NSString *modifiedString = [regex stringByReplacingMatchesInString:source options:0 range:NSMakeRange(0, [source length]) withTemplate:@""];
+
+    return modifiedString;
+}
+
+static int32_t IsIndexFound(int32_t fromBeginning, int32_t foundLocation, int32_t newLocation)
+{
+    // last index
+    if (!fromBeginning && foundLocation > newLocation)
+        return 1;
+    // first index
+    if (fromBeginning && foundLocation != -2 && foundLocation < newLocation)
+        return 1;
+    return 0;
+}
+
+/*
+Function: IndexOf
+Find detailed explanation how this function works in https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-hybrid-mode.md
+*/
+Range GlobalizationNative_IndexOfNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* lpTarget, int32_t cwTargetLength,
+                                        const uint16_t* lpSource, int32_t cwSourceLength, int32_t comparisonOptions, int32_t fromBeginning)
+{
+    assert(cwTargetLength >= 0);
+    Range result = {-1, 0};
+    NSStringCompareOptions options = ConvertFromCompareOptionsToNSStringCompareOptions(comparisonOptions);
+    
+    // in case mapping is not found
+    if (options == 0)
+        return result;
+    
+    NSString *searchString = [NSString stringWithCharacters: lpTarget length: cwTargetLength];
+    NSString *searchStrCleaned = RemoveWeightlessCharacters(searchString);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrCleaned = RemoveWeightlessCharacters(sourceString);
+
+    if (sourceStrCleaned.length == 0 || searchStrCleaned.length == 0)
     {
-        currentLocale = [NSLocale systemLocale];
+       result.location = fromBeginning ? 0 : sourceString.length;
+       return result;
     }
-    else
+
+    NSLocale *currentLocale = GetCurrentLocale(localeName, lNameLength);
+    NSString *searchStrPrecomposed = searchStrCleaned.precomposedStringWithCanonicalMapping;
+    NSString *sourceStrPrecomposed = sourceStrCleaned.precomposedStringWithCanonicalMapping;
+
+    // last index
+    if (!fromBeginning)
+        options |= NSBackwardsSearch;
+
+    // check if there is a possible match and return -1 if not
+    // doesn't matter which normalization form is used here
+    NSRange rangeOfReceiverToSearch = NSMakeRange(0, sourceStrPrecomposed.length);
+    NSRange containsRange = [sourceStrPrecomposed rangeOfString:searchStrPrecomposed
+                                                  options:options
+                                                  range:rangeOfReceiverToSearch
+                                                  locale:currentLocale];
+
+    if (containsRange.location == NSNotFound)
+        return result;
+
+    // in case search string is inside source string but we can't find the index return -2
+    result.location = -2;
```

Comment:
This may conflict with the `Debug.Assert(result.Location != -2);` in `CompareInfo.OSX.cs` because we might actually hit `-2` if we hit the unsupported mixed composition source string case. Perhaps we can use something different to indicate that this is a case that is currently unsupported.

Reply:
For this case changed return value to -3 and added exception.